### PR TITLE
ActiveRecord::Relation#async_pending? for introspecting async relations

### PR DESCRIPTION
**Pull request by joeldrapper**
Add `pending?` predicate to `ActiveRecord::Relation` that returns `true` if the relation is pending on the background thread pool; otherwise, returns `false`.

### Motivation / Background

This Pull Request has been created because there is currently no way to tell if an asynchronous `ActiveRecord::Relation` has actually loaded, since `loaded?` returns `true` even if the future is still pending.

An alternative approach might be to update the `loading?` predicate to return `false` when a future result is still pending. That would be a breaking change, and I don't know what else it might impact.

### Detail

This Pull Request adds a `pending?` predicate that returns `true` if the future result is pending; otherwise, returns `false`.

### Additional information

This change is based on [this discussion](https://discuss.rubyonrails.org/t/is-there-a-way-to-check-of-an-async-database-query-has-actually-loaded/82538).

I tried to write a test but wasn’t able to work out how to artificially slow an SQLite query to the point where it would still be pending by the time we check the predicate. There doesn’t seem to be a way to make an SQLite query sleep so let me know if you have any suggestions on how to do this.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by joeldrapper** [comment]
One option for the test is to use a different database adapter. Postgres, for example, has a `pg_sleep` function that would allow us to artificially slow a query enough to test this feature. Alternatively, we could stub the future.

**Review by eileencodes — CHANGES_REQUESTED** [review]
Thanks @joeldrapper can you write a test for this?

**Comment by joeldrapper** [comment]
Hey @eileencodes I had a look through the file you mentioned and while it looks like there's a helper to ensure an async query isn't pending, there doesn't seen to be a pattern here for getting an async query into a state where it's definitely pending.

The two options I can think of are

1. Use the Postgres adapter for this test and simulate a slow query with `select("pg_sleep(0.1)")`. (There doesn't seem to be sleep function in SQLite)
2. Stub the future so that it returns `true` when `pending?` is called

I’ll proceed with option 2 for now but let me know if you think there's a better way to do this.

**Comment by p8** [comment]
@joeldrapper Maybe the following comment is helpful for the test? https://github.com/rails/rails/pull/47503#issuecomment-1455109643

**Comment by joeldrapper** [comment]
Thanks @p8. I’m not sure how this applies. I think the stub is appropriate given that the futures `pending?` method can be tested independently and we can easily cover all three cases:

1. future doesn't exist
2. future exists and is truthy
3. future exists and is falsy

**Comment by matthewd** [comment]
What do we think about calling this something like `async_pending?`? It seems reasonable to expose, but I'm inclined to give it a more descriptive name to ensure that a future reader, glancing through the method list, doesn't see two semi-synonyms for "not currently loaded".

**Comment by p8** [comment]
> Thanks @p8. I’m not sure how this applies. 

https://github.com/rails/rails/blob/3cca0d5205329e28a9879739bb2d3ae3cff1b14f/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L71

The basic gist in pseudo code would be something like the following.
Although this might be overkill.

```ruby
    def test_pending_predicate_with_async_query
        $event = Concurrent::Event.new

        waiting_result = Class.new(ActiveRecord::FutureResult) do
        private
          def exec_query(*, **)
            $event.wait
            super
          end
      end
    
      deferred_posts = Post.where(author_id: 1)
      deferred_posts.instance_variable_set(:@future_result, waiting_result.new)

     assert_predicate deferred_posts, :pending?
     $event.set
     assert_not_predicate deferred_posts, :pending?
   end
```

Not sure if `Concurrent::Event` is even necessary.
But it might be nice to use dependency injection for the `FutureResult` class here:
https://github.com/rails/rails/blob/3cca0d5205329e28a9879739bb2d3ae3cff1b14f/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L71

**Comment by joeldrapper** [comment]
@matthewd I think that makes sense. I’ve renamed the method to `async_pending?` in the latest commit. I’ll rebase if this is approved.

File: activerecord/lib/active_record/relation.rb [hunk 1]
```diff
@@ -693,6 +693,12 @@ def scheduled?
       !!@future_result
     end
 
+    # Returns <tt>true</tt> if the relation is pending on the background
+    # thread pool; otherwise, returns <tt>false</tt>.
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation.rb:697]:
```suggestion
    # thread pool.
```

**Comment by joeldrapper** [comment]
I’ve rebased this PR with the suggested changes. Also, David Mills [suggested](https://discuss.rubyonrails.org/t/is-there-a-way-to-check-of-an-async-database-query-has-actually-loaded/82538/4?u=joeldrapper) an alternative name for the inverse of this as `ready?`

> What I would want to know is if the data fetched by the query was ready for use. To my mind `ready?` captures the concept as a synchronously loaded relation is _ready_ when it is loaded, but an asynchronously loaded relation is ready when the result is available.

If the maintainers agree this is a better name, I’m happy to make the necessary changes. Otherwise, I think this is ready for another review. Thanks!

**Comment by matthewd** [comment]
For now I think I prefer this: it's a slightly lower-level concept than (essentially) "will `to_a` block?", but that feels like it fits with the current UX of `load_async`. 

In a future where, among other things, we felt comfortable automatically shunting queries into async, I might imagine that most potential callers of a `ready?`, your use case included, might actually want a single method that will start an async load if one isn't already running, _then_ return whether the result is already available. If async is safe and free, and the result isn't loaded yet, ISTM you should take the time to do your flush even if you have to initiate the query.

I think your test needs an `unless in_memory_db?` around the method (or to migrate into the existing one).

If you can also fix the commit message to reflect the new method name, LGTM.

I did miss the existing adjacent `scheduled?`. I think my "pending sounds too ambiguous to someone not thinking of async" argument still holds [and doesn't apply to 'scheduled']. 🤔

**Comment by joeldrapper** [comment]
@matthewd makes sense. I updated the commit message and added that condition to the tests.

File: activerecord/test/cases/relation/load_async_test.rb [hunk 2]
```diff
@@ -41,6 +41,28 @@ def test_scheduled?
       assert_not_predicate deferred_posts, :scheduled?
     end
 
+    def test_async_pending_predicate_with_async_query
+      unless in_memory_db?
```
Comment (byroot) [inline activerecord/test/cases/relation/load_async_test.rb:45]: If the test doesn't apply to `in_memory_db?` the unless should be outside the `def`, like other tests in this file.

**Review by byroot — COMMENTED** [review]
I'm not sure I quite understand the use case here.

But assuming I can see one, I think we need to be careful about the semantic we want to expose.

- When you call `load_async`, the relation become `scheduled?`.
- When a background thread pick it up and executes it, it is still `scheduled?` but is also `pending?`
- Finally, when the background thread is done, it's now `loaded?`

I'm not sure I see in which case you need `pending?` where `scheduled?` wouldn't work better.

**Comment by joeldrapper** [comment]
I thought that was how `loaded?` works, but it actually becomes `loaded?` immediately when scheduled and there isn't a way to tell if it's actually finished loading in the background.

**Comment by byroot** [comment]
> there isn't a way to tell if it's actually finished loading in the background.

Yes there is. `loaded? && !scheduled?`.

**Comment by joeldrapper** [comment]
> Yes there is. `loaded?` && !scheduled?`.

That doesn't seem to be how `scheduled?` works.

Currently:

1. When you call `load_async`, the relation becomes `scheduled?` and `loaded?`. [Apparently](https://discuss.rubyonrails.org/t/is-there-a-way-to-check-of-an-async-database-query-has-actually-loaded/82538/3?u=joeldrapper), loaded here means “a query was sent to the database for a result”.
2. When the future result completes, the relation is still `scheduled?` and `loaded?`. It never becomes not `scheduled?`.

What I’m proposing is:

1. When you call `load_async`, the relation becomes `scheduled?` and `loaded?` and `pending?`
2. When the future result is ready, the relation becomes not `pending?`

File: activerecord/lib/active_record/relation.rb [hunk 3]
```diff
@@ -693,6 +693,12 @@ def scheduled?
       !!@future_result
     end
 
+    # Returns <tt>true</tt> if the relation is pending on the background
+    # thread pool.
+    def async_pending?
```
[the remark this answers is not in the document]
Comment (joeldrapper) [reply activerecord/lib/active_record/relation.rb:698]: I originally proposed `pending?`. See https://github.com/rails/rails/pull/47751#issuecomment-1482112239

Happy to change it back if that's the consensus.

File: activerecord/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,35 @@
...
+
+    ```ruby
+    @posts = Post.all.load_async
+    ```
+
+    If you wanted to define a function that would flush a buffer if it was waiting on an asynchronous
+    query, you could check the status of the query like this:
+
+    ```ruby
+    def await(query)
```
[the remark this answers is not in the document]
Comment (joeldrapper) [reply activerecord/CHANGELOG.md:19]: When given an asynchronous query and chained like the example below, it does await the query. I can remove the example, but I couldn't think of a better example.

**Comment by byroot** [comment]
> 2\. It never becomes not `scheduled?`.

Yes it does: https://github.com/rails/rails/blob/9f49959fee522f2f38a07eb07a23830b99db30f6/activerecord/lib/active_record/relation.rb#L924-L926

Might be a bit too late for your use case though, but we could change `scheduled?` to check that too and "finalize" the loading if the background query is completed.

File: activerecord/CHANGELOG.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (byroot) [reply activerecord/CHANGELOG.md:19]: What "await the query" in the example is the `.each`, not the `await()`, so yeah, I think this is misleading.

Anyway, I don't think we need a `pending?` method, it's redudant with `scheduled?`, we may just need to improve `scheduled?` a bit.

**Comment by joeldrapper** [comment]
I see so that happens after looking at (and waiting on) the result? I can try to change `scheduled?` so it starts returning `false` once the background query is no longer `pending?`.

**Comment by byroot** [comment]
> I can try to change `scheduled?` so it starts returning `false` once the background query is no longer `pending?`.

Yes.

**Comment by matthewd** [comment]
I had wondered about that, but then realised it would mean changing the condition in `load` to look elsewhere (because it cares about whether the records have been materialized, not just whether the query result has returned), so that seemed like it would still leave us needing the same set of methods no matter how we shuffle the names.

**Comment by byroot** [comment]
> so that seemed like it would still leave us needing the same set of methods no matter how we shuffle the names.

Yes, but the distinction between `scheduled` and `pending` would remain entirely a private concern, and wouldn't be exposed as a method (or would be private).

I see no use for publishing that state difference (but not opposed to doing it if someone comes up with a concrete use case).

**Comment by bensheldon** [comment]
I just wanted to chime in with a perspective from Concurrent Ruby, which makes a [distinction between `pending` and `processing`](https://github.com/ruby-concurrency/concurrent-ruby/blob/e7ff945661a93a2e094b3259f23ccb1b3eea9bcf/lib/concurrent-ruby/concurrent/promise.rb#L35-L38) in Futures/Promises/IVar, which like `ActiveRecord::FutureResult` will process either in a background thread _or the foreground thread_ if it is resolved before a background thread picks it up. (apologies if I'm misunderstanding FutureResult).  

This is a pedantic distinction, but it seems relevant if the intention around the `await` example is to orchestrate across threads. The counter-scenario is that it's possible that resolving the `pending` FutureResult(s) on the current thread is faster than performing the `await`ed extra-stuff and then resolving the Futures; it's possible this could _all_ take place on the foreground thread if just looking at `pending` and the Executor is otherwise busy.

If the intention is to know when it is actually executing on a background thread, my suggestion would be to introduce the idea of `processing` that would be set to true within `FutureResult#execute_or_skip` which is the method that is run on the background thread.

https://github.com/rails/rails/blob/89bd41201a060e8e6ee39591430d136e9e012c34/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb#L500-L503

**Comment by byroot** [comment]
@bensheldon concurrent-ruby operates at a much lower level, and yes in this case distinguishing between "queued" and "executing" is important. We do even use this distinction internally when iterating on a relation to know if we should perform the query or wait on the background thread.

What I'm saying is that I don't see an use case for that distinction at the Relation layer.

**Comment by bensheldon** [comment]
Thanks @byroot. I think I've convinced myself that the `await` use case is valid and functional (though still not optimal without looking into the state of the future itself, which I agree isn't appropriate for the Relation interface).

Changing `Relation#scheduled?` to return false after the query result is returned would functionally [introduce a seam between](https://github.com/bensheldon/rails/blob/03a1da9c34f6a86d2117fe87f1e60b0259d32337/activerecord/lib/active_record/relation.rb#L921-L932) the query result returning (IO) and instantiating the records (Ruby). Though I still think it's incorrect to imply via documentation that _it's on a background thread_... it's more like: take advantage of the situation on this thread while the query _might/could_ be on a different thread.

**Comment by matthewd** [comment]
@bensheldon I'm not sure whether it helps / is relevant, but for some context I'll note that Relation is not thread-safe (other than internally, relative to its own async behaviour), so no thread other than its "owner" should be poking at it, and I think that simplifies some of the anticipated requirements around a raw `await`.

@byroot I do think there's a plausibly-useful distinction between "if I take a while here, I won't be blocking the query from getting underway" ("has `load_async` been called?"), vs "the result would arrive faster if I called `load` right now".

Specifically, if the query has been async-ified but has not yet been started (because the async executor is busy), then ISTM you'd generally want to treat that the same as a not-loaded non-async relation:

* You know there's going to be delay before the result is available (so maybe you _would_ choose to flush output in the inspiring use case here), 
* but you also know that the sooner you block waiting on the result, the sooner the query will be started (so maybe you would choose _not_ to do anything that might even marginally slow that down).

Perhaps an API middle-ish ground would be a `scheduled?` that returns nil when it's not scheduled, or a more-specific-state symbol if it is? (`:future` vs `:running`, or whatever) 🤔

**Comment by byroot** [comment]
> plausibly-useful

Yes, but I can't really think of a practical use case for it.

**Comment by joeldrapper** [comment]
Closing this as it’s out of date now and there’s no consensus on the desired API. Happy to take another look if this comes up again.

**Comment by joeldrapper** [comment]
Posting this as a practical use case for this feature. https://x.com/joeldrapper/status/1838715247181377650

If we can agree on a name, I’m happy to re-open this PR and get it up to date.

**Pull request closed without merging**